dot/sync/chain_sync.go

```diff
@@ -402,23 +435,49 @@ func (cs *chainSync) sync() {
 
 			cs.tryDispatchWorker(worker)
 		case <-ticker.C:
-			worker, err := cs.handler.handleTick()
+			cs.maybeSwitchMode()
+
+			workers, err := cs.handler.handleTick()
 			if err != nil {
 				logger.Error("failed to handle tick", "error", err)
 				continue
 			}
 
-			if worker == nil {
+			if workers == nil {
 				continue
 			}
 
-			cs.tryDispatchWorker(worker)
+			for _, worker := range workers {
+				cs.tryDispatchWorker(worker)
+			}
+		case fin := <-cs.finalisedCh:
+			// on finalised block, call pendingBlocks.removeLowerBlocks() to remove blocks on
+			// invalid forks from the pending blocks set
+			cs.pendingBlocks.removeLowerBlocks(fin.Header.Number)
 		case <-cs.ctx.Done():
 			return
 		}
 	}
 }
 
+func (cs *chainSync) maybeSwitchMode() {
+	head, err := cs.blockState.BestBlockHeader()
+	if err != nil {
+		logger.Error("failed to get best block header", "error", err)
+		return
+	}
+
+	target := cs.getTarget()
+	if big.NewInt(0).Add(head.Number, big.NewInt(maxResponseSize)).Cmp(target) == -1 {
+		// we are 128 blocks or more behind the target, switch to bootstrap mode
+		cs.setMode(bootstrap)
+	} else if head.Number.Cmp(target) >= 0 {
```

Comment:
shouldn't this be an `else`? Is there another unaccounted for case?

Reply:
yeah, the `else` case is if the head is between `(target-128, target)`. in that case we don't want to switch modes